docs/developer-docs/security/formal-verification.md

```diff
@@ -0,0 +1,31 @@
+---
+sidebar_position: 4
+sidebar_label: Formal verification
+---
+# Formal verification
+
+Formal verification is the highest form of quality assurance for software. Given a specification of what the system should do, formal verification tools check whether this specification is satisfied by a model of the system. The unique advantage of formal verification is that it can not only find bugs but also formally **prove** their absence – including the absence of security bugs. This goes beyond what testing or manual audits can achieve.
+
+The proof is always relative to the model and the specification. Any simplifications and assumptions in the model, or omissions in the specification, may hide bugs and attacks. On the other hand, verification can require a lot of effort, and model simplifications can make it significantly easier.
+
+For a concrete example, when verifying the ckBTC minter canister, the DFINITY team used models that exclude the possibility of calling the `update_balance` canister method more than 2 times concurrently. This potentially misses attacks that require 3 or more concurrent calls to `update_balance` to trigger a bug. But we considered such bugs highly unlikely and, in return, we were able to run a fully automatic verification process, which was much cheaper than other verification methods.
+
+There are many existing formal verification tools. While none of them take into account the specifics of ICP yet, many of them are general enough that they can be applied to canisters. In particular, the DFINITY team has made good use of the TLA+ toolkit to eliminate reentrancy bugs, a type of concurrency bug where one method of a particular canister is called while another one is still executing.
```

Comment:
"eliminate reentrancy" bugs seems like too strong of a claim. "Eliminate" makes it sound like we proved the correctness with regard to reentrancy in all DFINITY code. I think something like "reveal" would be more appropriate.